packages/plugin-data/src/api/data/tree/exportApi.ts

```diff
@@ -0,0 +1,607 @@
+      if (err.name !== 'EEXIST') {
+        throw err;
+      }
+    }
+  }
+
+  // Process query results generating SObject Tree format
+  private async processQueryResults(recordList: any): Promise<any> {
+    await this.recordObjectTypes(recordList);
+
+    const { plan, query } = this.config;
+
+    const processedRecordList = await this.processRecordList(recordList);
+    // log record count; warn if > 200 and !options.plan
+    const recordCount = getNumber(processedRecordList, 'records.length', 0);
+    this.logger.debug(messages.getMessage('dataExportRecordCount', [recordCount, query]));
+    if (recordCount > 200 && !plan) {
+      this.ux.warn(messages.getMessage('dataExportRecordCountWarning', [recordCount, query]));
+    }
+    return this.finalApplyRefs(processedRecordList.records);
+  }
+
+  // Register object types and type hierarchy for plan generation
+  private async recordObjectTypes(recordList: any): Promise<any> {
+    const records = recordList.records;
+
+    if (!records.length) {
+      // TODO: should be on the command
+      this.ux.log('Query returned no results');
+      return recordList;
+    }
+
+    // top level object type
+    this.objectTypeRegistry[records[0].attributes.type as string] = {
+      order: 0,
+      type: records[0].attributes.type,
+      saveRefs: true, // post-save, record reference-to-id to be used for child reference resolution
+      resolveRefs: false, // pre-save, don't resolve relationship references to parent ids (from previous save)
+    };
+
+    records.forEach((record: any) => {
+      Object.keys(record).map((key) => {
+        const value = record[key];
+        if (value && getNumber(value, 'records.length')) {
+          const firstRec = value.records[0];
+          const type = getString(firstRec, 'attributes.type');
+          // found a related object, add to map
+          if (type && !this.objectTypeRegistry[type]) {
+            this.objectTypeRegistry[type] = {
+              order: 1,
+              type,
+              saveRefs: false, // assume child records will not be parents (may be changed later)
+              resolveRefs: true, // resolve relationship references to parent ids
+            };
+          }
+        }
+        return key;
+      });
+    });
+
+    // pre-load object metadata
+    const promises = Object.keys(this.objectTypeRegistry).map((key) => this.loadMetadata(key));
+    return Promise.all(promises).then(() => recordList);
+  }
+
+  private async processRecordList(recordList: any, parentRef?: any): Promise<any> {
+    // holds transformed sobject tree
+    const sobjectTree = { records: [] };
+
+    // visit each record in the list
+    const processRecordsFn = (record: any) => (): Promise<any> => this.processRecords(parentRef, record, sobjectTree);
+    const recordFns = recordList.records.map((record: any) => processRecordsFn(record));
+
+    // TODO: do we really need sequentialExecute?  Can't we just use Promise.all()?
```

Comment:
can't we? I feel like this is the time to fix all of these TODOs and make the plugin as pristine as possible